File: Matrices.js
const readline = require('readline/promises');
const { stdin: input, stdout: output } = require('process');

const rl = readline.createInterface({ input, output });

// Array de objetos. Ejemplo: [{ nombre: "Juan", materias: { "Matemáticas": 9, "Historia": 8 } }]
let alumnos = [];

const calcularPromedio = (materias) => {
    const notas = Object.values(materias);
    if (notas.length === 0) return 0;
    const total = notas.reduce((suma, nota) => suma + nota, 0);
    return total / notas.length;
};

const buscarAlumno = (nombre) => 
    alumnos.find(a => a.nombre.toLowerCase() === nombre.toLowerCase());

function verAlumnos() {
    if (alumnos.length === 0) {
        return console.log("\nNo hay alumnos registrados.");
    }

    const alumnosOrdenados = [...alumnos].sort((a, b) => 
        calcularPromedio(b.materias) - calcularPromedio(a.materias)
    );

    console.log("\n--- Lista de Alumnos ---");
    alumnosOrdenados.forEach(({ nombre, materias }) => {
        const promedio = calcularPromedio(materias);
        console.log(`\nAlumno: ${nombre} | Promedio: ${promedio.toFixed(2)}`);
        
        Object.entries(materias).forEach(([materia, nota]) => {
            console.log(`  - ${materia}: ${nota}`);
        });
    });

    const mejor = alumnosOrdenados[0];
    console.log(`\n🏆 Mejor alumno: ${mejor.nombre} con un promedio de ${calcularPromedio(mejor.materias).toFixed(2)}`);
}

async function agregarAlumno() {
    const nombre = await rl.question("\nIngrese el nombre del alumno: ");
    
    if (buscarAlumno(nombre)) {
        console.log("El alumno ya está registrado. Ve a la opción 3 para modificar sus notas.");
    } else {
        // Guardamos un objeto en lugar de un array
        alumnos.push({ nombre, materias: {} });
        console.log(`Alumno ${nombre} agregado con éxito.`);
    }
}

async function gestionarNotas() {
    const nombre = await rl.question("\nIngrese el nombre del alumno: ");
    const alumno = buscarAlumno(nombre);

    if (!alumno) {
        return console.log("El alumno no existe. Por favor, agréguelo primero (Opción 2).");
    }

    const materia = await rl.question("Ingrese el nombre de la materia: ");
    const nota = parseFloat(await rl.question("Ingrese la nota: "));

    if (isNaN(nota)) {
        return console.log("La nota debe ser un número válido.");
    }

    if (alumno.materias[materia] !== undefined) {
        console.log(`Modificando nota de ${materia}: de ${alumno.materias[materia]} a ${nota}`);
    } else {
        console.log(`Materia ${materia} agregada con la nota ${nota}.`);
    }
    
    alumno.materias[materia] = nota;
}

async function main() {
    let salir = false;
    while (!salir) {
        console.log("Ver alumnos y estadísticas");
        console.log("Agregar alumno");
        console.log("Agregar o modificar notas");
        console.log("Salir");
        const opcion = await rl.question("Elige una opción");

        switch (opcion) {
            case '1':
                verAlumnos();
                break;
            case '2':
                await agregarAlumno();
                break;
            case '3':
                await gestionarNotas();
                break;
            case '4':
                console.log("Saliendo del sistema...");
                salir = true;
                rl.close();
                break;
            default:
                console.log("Opción no válida. Intenta de nuevo.");
        }
    }
}

// Ejecutar el programa
main();

File: Matrices.py
def calcular_promedio(materias):
    if not materias:
        return 0
    total = sum(materias.values())
    return total / len(materias)

def buscar_alumno(alumnos, nombre):
    for alumno in alumnos:
        if alumno["nombre"].lower() == nombre.lower():
            return alumno
    return None

def obtener_promedio_para_ordenar(alumno):
    return calcular_promedio(alumno["materias"])

def ver_alumnos(alumnos):
    if not alumnos:
        print("\nNo hay alumnos registrados.")
        return

    alumnos_ordenados = sorted(alumnos, key=obtener_promedio_para_ordenar, reverse=True)
    
    print("\n--- Lista de Alumnos ---")
    for alumno in alumnos_ordenados:
        nombre = alumno["nombre"]
        materias = alumno["materias"]
        promedio = calcular_promedio(materias)
        
        print(f"\nAlumno: {nombre} | Promedio: {promedio:.2f}")
        
        for materia, nota in materias.items():
            print(f"  - {materia}: {nota}")
            
    if alumnos_ordenados:
        mejor = alumnos_ordenados[0]
        promedio_mejor = calcular_promedio(mejor["materias"])
        print(f"\n🏆 Mejor alumno: {mejor['nombre']} con un promedio de {promedio_mejor:.2f}")

def agregar_alumno(alumnos):
    nombre = input("\nIngrese el nombre del alumno: ")
    if buscar_alumno(alumnos, nombre):
        print("El alumno ya está registrado. Ve a la opción 3 para modificar sus notas.")
    else:
        alumnos.append({"nombre": nombre, "materias": {}})
        print(f"Alumno {nombre} agregado con éxito.")

def gestionar_notas(alumnos):
    nombre = input("\nIngrese el nombre del alumno: ")
    alumno = buscar_alumno(alumnos, nombre)
    
    if not alumno:
        print("El alumno no existe. Agréguelo primero (Opción 2).")
        return
        
    materia = input("Ingrese el nombre de la materia: ")
    try:
        nota = float(input("Ingrese la nota: "))
    except ValueError:
        print("La nota debe ser un número válido.")
        return

    materias_del_alumno = alumno["materias"]
    
    if materia in materias_del_alumno:
        print(f"Modificando nota de {materia}: de {materias_del_alumno[materia]} a {nota}")
    else:
        print(f"Materia {materia} agregada con la nota {nota}.")
        
    materias_del_alumno[materia] = nota

def main():
    alumnos = []
    
    while True:
        print("Ver alumnos y estadísticas")
        print("Agregar alumno")
        print("Agregar o modificar notas")
        print("Salir")
        
        opcion = input("Elige una opción")
        
        if opcion == '1':
            ver_alumnos(alumnos)
        elif opcion == '2':
            agregar_alumno(alumnos)
        elif opcion == '3':
            gestionar_notas(alumnos)
        elif opcion == '4':
            print("Saliendo del sistema...")
            break
        else:
            print("Opción no válida. Intenta de nuevo.")

if __name__ == "__main__":
    main()